modules/nextflow/src/main/groovy/nextflow/file/FilePorter.groovy

```diff
@@ -156,16 +165,24 @@ class FilePorter {
         return result
     }
 
+    @Canonical
+    static class FileCopy {
+        final Path source
+        final Path target
+    }
+
     /**
      * Models a batch (collection) of foreign files that need to be transferred to
      * the process staging are co-located with the work directory
      */
     static class Batch  {
 
+        final private FilePorter owner
+
         /**
          * Holds the list of foreign files to be transferred
          */
-        private List<Path> foreignPaths = new ArrayList<>()
+        private List<FileCopy> foreignPaths = new ArrayList<>(100)
```

Comment:
Instead of the file of source paths use a list of pairs. Each pair represent the source file location and the target location of the staging.